### DISCUSSION TITLE
Software not writing to EEPROM

### DISCUSSION THREAD

**Discussion by sirquackiii** [TomNisbet/TommyPROM#31] [Q&A] [answered]
I'm currently working on the Ben Eater 6502 project, and I'm at the point where I have to write to the EEPROM. I edited the code for the Ben Eater unlock sketch to write 0xea to some addresses
![image](https://user-images.githubusercontent.com/77940766/124526456-b9bf2e80-ddb7-11eb-8ec5-d667950aa984.png)
but none of it seemed to actually write to the chip and is still filled with 0xff.
![image](https://user-images.githubusercontent.com/77940766/124526541-0b67b900-ddb8-11eb-96b4-1f92dd4a90f7.png)
What's wrong with this? I checked my wiring and everything _should_ be correct, and in the code it already disables SDP, so I'm not sure what's happening.

Thanks.

**Comment by TomNisbet** [comment, marked as answer]
It appears that your serial communication problem is solved.  What was the issue?

Given that all of the data is coming back as FF, I suspect that the chip is still locked.  Once unlocked, it is ridiculously easy to write _something_ to the chip, even by accident.  Ben's design ties the WE line to the D13 line of the Arduino and that is used to flash the LED at boot, so you could even see stray bytes written by that if the chip were unlocked.

The problem is unlikely to be timing because the sketch unlock code is well within the specs for all of the chips I've seen.  It is possible that you have an address or data line swapped, unconnected, or shorted.  That would mean that the chip wouldn't see the correct sequence to disable SDP.  A few things to try:

- Send me a hi-res picture (or two) of your board so I can see the wiring.
- What is the exact part number you are using for the chip?  For example, AT28C256-15PC is an Atmel chip.
- What kind of Arduino are you using?  This sketch is made specifically for the Nano or Uno and will fail on others that map the IO ports differently.

The TommyPROM documentation has a section on hardware debugging.  That code isn't written for the Ben Eater hardware, but the ideas are the same: remove the EEPROM chip and set know address and data values and probe the lines to make sure that the correct signals are present.

**Reply by sirquackiii** [reply]
I rewired my programmer, and found that I had twisted some address wires, and now I believe it's working!
![image](https://user-images.githubusercontent.com/77940766/124832698-e8611480-df31-11eb-8c99-dc804f284a05.png)

**Reply by sirquackiii** [reply]
And about the serial communication problem, it wasn't a problem, I believe I connected to the serial port a bit too late on Tera Term, so it would show nothing. I decided to use the Arduino IDEs serial monitor, and it works fine.

**Reply by TomNisbet** [reply]
In a way, you are lucky that the chip was locked and the address problem prevented the unlock.  Otherwise, your programmer would have used the same wrong addresses for reading and writing and it would have looked like it was working.  But when you put the EEPROM into your other circuit all of the data would be mixed up in the wrong addresses and it wouldn't have worked.  That would not be an easy debug!  Glad it is working for you.
